### PR TITLE
[SYCL] Check if the underlying buffer type is device copyable

### PR DESCRIPTION
**Pull request by psamolysov-intel**
The implementation should diagnose an error if the underlying type "T"
of a buffer is not device copyable. The spec says that this is
a requirement for all buffers. It does not make any exception for
buffers that are used only on the host:

"The underlying data type of a buffer T must be device copyable as
defined in Section 3.13.1."

(https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:buffers)

### CODE REVIEW

**Comment by psamolysov-intel** [comment]
Since the `sycl::vec` class is trivially copyable only on device, I've added a workaround to mark `sycl::vec` as device copyable on host if and only if the underlying type is also device copyable.

**Review by vladimirlaz — DISMISSED** [review]

File: sycl/include/CL/sycl/types.hpp [hunk 1]
```diff
@@ -2420,6 +2420,15 @@ struct is_device_copyable<sycl::marray<T, N>,
                           std::enable_if_t<is_device_copyable<T>::value>>
     : std::true_type {};
 
+// vec is device copyable on host if element type is device copyable,
+// on device vec is trivially copyable and therefore is device copyable too.
+#ifndef __SYCL_DEVICE_ONLY__
+template <typename T, int N>
+struct is_device_copyable<sycl::vec<T, N>,
+                          std::enable_if_t<is_device_copyable<T>::value>>
```
[the remark this answers is not in the document]
Comment (alexbatashev) [reply sycl/include/CL/sycl/types.hpp:2428]: +1 on this comment. While this file is too crowded with templates, that removing this one clause will not make any difference at all, it is still a good idea not to produce unnecessary template instantiations.

File: sycl/test/basic_tests/buffer/buffer_for_not_device_copyable.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// RUN: %clangxx -fsycl -fsyntax-only -Xclang -verify %s -Xclang -verify-ignore-unexpected=note,warning
+
+#include <sycl/sycl.hpp>
+#include <string>
+
+using namespace cl::sycl;
```
Comment (alexbatashev) [inline sycl/test/basic_tests/buffer/buffer_for_not_device_copyable.cpp:6]:
```suggestion
using namespace sycl;
```

File: sycl/include/CL/sycl/types.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (psamolysov-intel) [reply sycl/include/CL/sycl/types.hpp]: @gmlueck @alexbatashev Thanks for the suggestion. I've removed the check for the underlying type `T`.

File: sycl/test/basic_tests/buffer/buffer_for_not_device_copyable.cpp (hunk 2 above)
Comment (psamolysov-intel) [reply sycl/test/basic_tests/buffer/buffer_for_not_device_copyable.cpp]: @alexbatashev Thank you, the namespace is changed to `sycl::`.

**Review by alexbatashev — DISMISSED** [review]
lgtm

**Review by vladimirlaz — APPROVED** [review]

**Review by gmlueck — APPROVED** [review]

**Review by vladimirlaz — APPROVED** [review]

**Pull request merged**